### ISSUE TITLE
Support for CD+G

### ISSUE THREAD

**Issue by tjanas** [MiSTer-devel/TurboGrafx16_MiSTer#189]
Real TurboGrafx CD hardware supports CD+G discs.
This is a feature request to support ccd/img/sub CD+G disc images.

**Comment by Left-Empty** [comment]
I don't know this is the place to discuss requests (the forum is a better place probably), but in my experience (which were only with a couple of such CD+G discs), my PC Engine Duo didn't support them. They worked fine on my Mega CD though.
I seem to recall some sort of karaoke expansion being sold, so maybe it was required to unlock the proper functionnality?

**Comment by dshadoff** [comment]
The original machine does support CD+G, but this core does not currently support that.
I think filing this issue was not necessary... the feature was never identified as "not to be implemented", nor is it something the core developers were unaware of; it simply hasn't been implemented yet.

It has been on my mind, and I might try to implement it if I get a sufficient block of free time to do so.

**Comment by tjanas** [comment]
Please feel free to close the issue, I wasn’t sure if it was on the radar.

**Comment by sorgelig** [comment]
This feature is on its own. It has no more than browse the pictures. It was actual back in 90x, but who will use it now?

**Comment by dshadoff** [comment]
While people may use the feature, it is also already available in other cores (for example the Mega CD).

**Comment by tjanas** [comment]
CD+G already works in the [MegaCD_MiSTer](https://github.com/MiSTer-devel/MegaCD_MiSTer) core?

**Comment by dshadoff** [comment]
Yes it does.  Zakk added it in summer of 2021 (it was already fleshed-in on the core, but the Main_MiSTer changes were committed then).  I had helped to test at that time.

**Comment by dshadoff** [comment]
Just leaving a note here - not sure if this is implemented yet, but there is an IRQ2 fired approximately every 135 microseconds once the user enters the "GRAPHICS" screen on the CD System Card; this is apparently related to CD+G.

see discussion in #220 for how to read the following logic trace
- Test starts with disc loaded, and at the CD player screen
- When "GRAPHICS" screen is started, the IRQ2 starts getting fired frequently, even before PLAY is chosen
- IRQ stops firing when you return to the main CD player screen.

[DSLogic_PCE_IFU_CDR_CDG_on_GFX_scr.zip](https://github.com/user-attachments/files/19761432/DSLogic_PCE_IFU_CDR_CDG_on_GFX_scr.zip)

**Comment by dshadoff** [comment]
In a (private workspace) test update, I have confirmed that resending the "current" sector during PAUSE state makes the controls usable, and prevent the player from forcing a disc change.

However, pressing "STOP" apparently puts the player into "PAUSE" state, which may not be correct (I will look into this).

**Comment by dshadoff** [comment]
I sent a pull request to Main_MiSTer which should cover the remaining items (other than CHD).

1. Pause - This requires that subcode data is constantly sent, whether the head moves or not. If paused, repeat data is sent.
2. Stop - This is no longer an issue in this update
3. Playing a CD without subcode file, in the GRAPHICS player - This requires synthesizing subcode-Q timing data, which has been done and is working fine.

**Comment by real-amano** [comment]
Latest **MiSTer** unstable build: https://github.com/MiSTer-unstable-nightlies/Main_MiSTer/releases/download/unstable-builds/MiSTer_unstable_20250528_08fd2e
```[David Shadoff 2025-05-28 fd2e50e]
TGFX16: Fixes pause on CD+G

* Fixes pause on CD+G (TurboGrafx16_MiSTer#189)

Fix pause on GRAPHICS playback screen by continuously sending subcode data. Also, synthesize subcode data when subcode file not found, so that regular CDs don't have issues in that function.

* suppress debug msg

**Comment by tjanas** [comment]
> Yes it does. Zakk added it in summer of 2021 (it was already fleshed-in on the core, but the Main_MiSTer changes were committed then). I had helped to test at that time.

On a side note, I just tried my CD+G images with the latest SegaCD core release, and there are still [major graphical glitches.](https://github.com/MiSTer-devel/MegaCD_MiSTer/issues/101#issuecomment-1463137839)

**Comment by dshadoff** [comment]
OK, the complete feature is now accepted into the codebase, and is in the current "unstable" build (not in the official release, because this was committed after the recent release).

@tjanas : Could you verify that it is working, and close the issue ?  (or report back with any issues you find)

**Comment by tjanas** [comment]
Working great with [MiSTer_unstable_20250609_183a49](https://github.com/MiSTer-unstable-nightlies/Main_MiSTer/releases/download/unstable-builds/MiSTer_unstable_20250609_183a49) and [TurboGrafx16_unstable_20250526_056080.rbf](https://github.com/MiSTer-unstable-nightlies/TurboGrafx16_MiSTer/releases/download/unstable-builds/TurboGrafx16_unstable_20250526_056080.rbf)
